### PR TITLE
Adding a Four In A Row app (08Sep23)

### PR DESCRIPTION
**Pull request by drtonyr**
This replaces #446 which was FUBAR

### CODE REVIEW

File: apps/four_in_a_row.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+~~~~~~~~~~~~~~~~
+ 
+This is the classic two player board game called Four In A Row or Connect4.  You play against the computer.
+
+.. figure:: res/screenshots/FourInARowApp.png
+    :width: 179
+
+    Screenshot of Four In A Row
+
+There is in intro/menu screen which has very brief instructions, allows you to set the opponent level and gives some stats on the number of games you have won.  Touching the screen sets the level from 0 to 6 which corresponds to the number of lookahead plies.  Swiping down enters the main game.  On your turn a red square counter will appear on the top row.  Touch the screen to move to the desired column, optinally touch again if you don't like your choice then swipe down to commit to playing that column.  The computer will reply with a yellow counter after a delay (dependent on level).  Your aim is to get four in a row before the computer does.  At end of game swipe down to return to the intro/menu screen.
```
Comment (fgaz) [inline apps/four_in_a_row.py:14]: typos
optinally→optionally
plies→piles?

File: apps/four_in_a_row.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+This app is powered by a compact version of the Alpha Beta pruning algorithm.  For technical details see https://en.wikipedia.org/wiki/Connect_Four and kaggle.com/competitions/connectx/.  There isn't space for a transposition table in RAM, so MTDf isn't implmented, nevertheless it can play a challenging game.
```
Comment (fgaz) [inline apps/four_in_a_row.py:16]: > kaggle.com/competitions/connectx/

I think this needs https:// at the beginning to be interpreted as URL by the doc renderer

File: apps/four_in_a_row.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            upb = min(upb, bestv)
+            if bestv <= lob:
+                break
+
+    return -bestv, bestx
+
+def rndColumn():
+    return int(wasp.watch.rtc.uptime) % _NCOLUMN  # time used as RNG to save space
+
+# class FourinaRowApp(): # names are different between firmware and /flash
```
Comment (fgaz) [inline apps/four_in_a_row.py:98]: Is this still relevant?

**Review by fgaz — COMMENTED** [review]
I didn't look at the code yet, but I found a few typos

File: apps/four_in_a_row.py (hunk 1 above)
Comment (drtonyr) [reply apps/four_in_a_row.py]: Thanks a lot for looking at this.
I've fixed 3 of the 4 suggestions you've made.
The only one I didn't change is 'plies', it has this meaning https://en.wikipedia.org/wiki/Ply_(game_theory)#:~:text=plies.

**Review by daniel-thompson — APPROVED** [review]
I've been dragging my feet for a long time as a maintainer in wasp-os. I'm busy with other things, I only drop on occasionally and when I am here I only ever look at the pull-requests (I never have time to look at the Issues). I'd like to get to a position where I am no longer needed to get code landed in wasp-os since having to wait a month for everything isn't a great contributor experience. For that reason I've invited you to join the project on github. That should turn on the "Merge this code" button for you in github (although it shouldn't appear until someone gives the PR a code review). If you'd like to accept then great... and I'd encourage you to get involved in other code reviews since that will help keep the project alive!

However, whether you accept or not, I think this can be merged. I confess I haven't run it but the code looks good... and if there are any problems they won't affect anything except this app ;-) .

**Comment by drtonyr** [comment]
Thank you for the invitation, it's an honour to accept.

I think I know my strengths and weaknesses.  I love wasp-os because it's similar to projects of my youth (I had wrote one of the first 32k ROMs for the BBC micro and was a developer for Sinclair's Z88).  I also know that I'm weak in git workflows but there are many here who can quickly get me up to speed with that.

There's lots I'd like to do - thanks again.

**Pull request merged**